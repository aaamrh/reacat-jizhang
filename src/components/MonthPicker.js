/* eslint-disable jsx-a11y/anchor-is-valid */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { range } from '../utils'



class MonthPicker extends Component {
  constructor(props){
    super(props)
    this.state = {
      isOpen: false,
      selectedYear: this.props.year,
      selectedMonth: this.props.month,
    }
  }

  componentDidMount(){
    document.addEventListener('click', this.handleClick, false)
  }
  componentWillUnmount() {
    document.removeEventListener('click', this.handleClick, false)
    console.log('will unmount')
  }
  
  handleClick = (event) => {
    // GET
    console.log('this.node', this.node)
    if (this.node.contains(event.target)) {
      return;
    }
    this.setState({
      isOpen: false,
    })
  }

  toggleDropDown = (ev) => {
    ev.preventDefault()
    this.setState({
      isOpen: !this.state.isOpen
    })
  }

  selectYear = (ev, yearNum) => {
    ev.preventDefault();
    this.setState({
      selectedYear: yearNum
    })
  }

  selectMonth = (ev, monthNum) => {
    ev.preventDefault()
    this.setState({
      isOpen: !this.state.isOpen,
      selectedMonth: monthNum
    })

    this.props.onChange( this.state.selectedYear, monthNum )
  }

  render() {

    const { year, month } = this.props

    const { selectedYear, selectedMonth } = this.state

    const monthRange = range(12, 1);

    const yearRange = range(9, -4).map( number => number + year )

    return (
      // GET ref, 获取DOM节点
      <div className="dropdown month-picker-component" ref={(ref) => { this.node = ref }}> 
        <h4>选择月份</h4>
        <button 
          className="btn btn-secondary dropdown-toggle" 
          onClick={ this.toggleDropDown }
        >
          { `${ selectedYear } 年 ${ selectedMonth } 月` }
        </button>

        { this.state.isOpen && 
          <div className="dropdown-menu" style={{display: 'block'}}>
            <div className="row">
              <div className="col border-right">
                { yearRange.map( (yearNum, index) => <a key={index} href='#'
                    className={ (yearNum === selectedYear) ? 'dropdown-item active' : 'dropdown-item'}
                    onClick = { (ev) => { this.selectYear(ev, yearNum) } }
                  >
                    { yearNum } 年
                  </a> ) 
                }
              </div>
              <div className="col">
              { monthRange.map( (monthNum, index) => <a key={index} href='#'
                className={ (monthNum === selectedMonth) ? 'dropdown-item active' : 'dropdown-item'}
                onClick = { (ev) => this.selectMonth(ev, monthNum) }
              >
                  { monthNum } 月
                </a> ) }
              </div>
            </div>
          </div>
        }
    </div>
    );
  }
}

export default MonthPicker;