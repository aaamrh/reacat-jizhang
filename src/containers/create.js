import React, { Component } from 'react';
import {withRouter} from 'react-router-dom'
import CategorySelect from '../components/CategorySelect';
import PriceForm from '../components/PriceForm';
import { testCategories } from '../testData'
import { Tabs, Tab } from '../components/Tabs';
import { TYPE_INCOME, TYPE_OUTCOME } from '../utils'
import { AppContext } from '../App'


const tabsText = [ TYPE_OUTCOME, TYPE_INCOME ]

class Create extends Component {
  constructor(props){
    super(props)

    this.state = {
      selectedTab: TYPE_OUTCOME,
      selectedCategory: null,
    }
  }

  cancelSubmit = () => {
    this.props.history.push('/')
  }

  selectCategory = (category) => {
    this.setState({
      selectedCategory: category
    })
  }

  tabChange = (index) => {
    this.setState({
      selectedTab: tabsText[index]
    })
  }

  render() {
    // const filterCategories = testCategories.filter( category => category.type===TYPE_OUTCOME )

    return (
      <AppContext.Consumer>
        { ( {state, actions} )=>{
          const { id } = this.props.match.params;
          const { items, categories } = state;
          const selectedTab = (id && items[id]) ? categories[ items[id].cid ].type : this.state.selectedTab;
          const selectedCategory = (id && items[id]) ? categories[ items[id].cid ] : this.state.selectedCategory

          const editItem = (id && items[id] ) ? items[id] : {}

          const filterCategories = Object.keys(categories)
          .filter( id=>categories[id].type === selectedTab)
          .map(id => categories[id])

          const submitForm = (data, isEditMode) => {
            if ( !isEditMode ){
              actions.createItem(data, selectedCategory.id)
            }else{
        
            }
            this.props.history.push('/')
          }
          
          return (
            
            <div className="create-page py-3 px-3 rounded mt-3" style={{ background: '#fff' }}>
              <Tabs activeIndex={0} onTabChange={ this.tabChange }>
                <Tab>支出</Tab>
                <Tab>收入</Tab>
              </Tabs>

              <CategorySelect categories={filterCategories} 
                onSelectCategory={ this.selectCategory }
                selectedCategory = { selectedCategory }
              />

              <PriceForm 
                onFormSubmit={ submitForm }
                onCancelSubmit={ this.cancelSubmit }
                item={editItem}
              />
            </div>
          )
        } }
      </AppContext.Consumer>

      // <div>
      //   create.js { this.props.match.params.id }
      //   {/* <CategorySelect categories={filterCategories} 
      //     onSelectCategory={this.selectCategory}
      //     selectedCategory={selectedCategory}
      //   /> */}
      //   <PriceForm 
      //     onFormSubmit={this.submitForm}
      //     onCancelSubmit={this.cancelSubmit}
      //     item={{}}
      //   />
      // </div>
    );
  }
}

export default withRouter ( Create );