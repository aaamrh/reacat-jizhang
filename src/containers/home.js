import React, { Component } from 'react';

import { LIST_VIEW, CHART_VIEW, parseToYearAndMonth } from '../utils'
import logo from '../logo.svg';

import PriceList from '../components/PriceList'
import ViewTab from '../components/ViewTab';
import TotalPrice from '../components/TotalPrice';
import MonthPicker from '../components/MonthPicker';
import CreateBtn from '../components/CreateBtn';


export const categories = {
  '0': {
    'id': 0,
    'name': '工资',
    'type': 'income'   
  },
  '1':{
    'id': 1,
    'name': '旅行',
    'type': 'outcome'   
  },
  '2':{
    'id': 2,
    'name': '零食',
    'type': 'outcome'   
  }

}

export const items = [
  {
    'id': 0,
    'title': '工资',
    'price': 5000,
    'date': '2020-12-10',
    'cid': 0
  },
  {
    'id': 1,
    'title': '去云南旅游',
    'price': 25,
    'date': '2021-1-11',
    'cid':1
  },
  {
    'id': 2,
    'title': '车票',
    'price': 115,
    'date': '2021-1-15',
    'cid':2
  }
]

const newItem = {
  'id': 4,
  'title': '新纪录',
  'price': 280,
  'date': '2021-1-18',
  'cid':1
}

class Home extends Component {
  constructor(props){
    super(props)
    this.state = {
      items,
      currentDate: parseToYearAndMonth(),
      tabView: LIST_VIEW
    }
  }


  changeView = (view) => {
    this.setState({
      tabView: view
    })
  }

  changeDate = (year, month) => {
    this.setState({
      currentDate: {year, month}
    })
  }

  createItem = () => {
    this.setState({
      items: [newItem, ...this.state.items]
    })
  }

  modifyItem = (modifiedItem) => {
    const modifiedItems = this.state.items.map( item => {
      if( item.id === modifiedItem.id ){
        return { ...item, title: '被修改' }
      }
      return item
    } )

    this.setState({
      items: modifiedItems
    })
  }
   
  deleteItem = (deletedItem) => {
    const filteredItems = this.state.items.filter( item => item.id !== deletedItem.id )

    this.setState({
      items: filteredItems
    })
  }

  render() {
    const {items, currentDate, tabView} = this.state;

    let totalIncome = 0, totalOutcome = 0;

    const itemsWithCategory = items.map( item => {
      item.category = categories[ item.cid ]
      return item
    } ).filter( item => item.date.includes(`${currentDate.year}-${currentDate.month}`) )

    items.forEach( item => {
      if( item.category.type === 'outcome' ){
        totalOutcome += item.price
      }else{
        totalIncome += item.price
      }
    } )

    return (
      <>
         <header className="App-header">
          <img src={logo} className="App-logo" alt="logo" />
        </header>

        <MonthPicker 
          year={currentDate.year}
          month={currentDate.month}
          onChange = { this.changeDate }
        />

        <TotalPrice 
          income = {totalIncome}
          outcome = {totalOutcome}
        />

        <CreateBtn onClick={ this.createItem } />

        <ViewTab 
          activeTab = { tabView }
          onTabChange = { this.changeView }
        />
        {
          tabView === LIST_VIEW && 
          <PriceList items={itemsWithCategory} 
            onModifyItem = {this.modifyItem}
            onDeleteItem = {this.deleteItem}
          />
        }
        {
          tabView === CHART_VIEW && 
          <h1>图表</h1>
        }
      </>
    );
  }
}

export default Home;