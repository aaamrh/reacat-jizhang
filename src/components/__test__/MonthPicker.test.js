import Enzyme, { render, mount } from 'enzyme'  
import MonthPicker from '../MonthPicker'
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { items } from '../../containers/home'


Enzyme.configure({ adapter: new Adapter() });

 
/** 
 * 犹豫涉及了一些只有DOM环境才存在的变量， 比如 event, document添加click事件
 * 因此需要使用mount
 * 
 * shallow:
 * shallow 返回组件的浅渲染，对官方shallow rendering 进行封装。
 * 浅渲染 作用就是：它仅仅会渲染至虚拟dom，不会返回真实的dom节点，对测试性能有极大的提升。
 * shallow只渲染当前组件，只能能对当前组件做断言
 * 
 * mount :
 * mount 方法用于将React组件加载为真实DOM节点。mount会渲染当前组件以及所有子组件
 * 
 * render: 
 * render 采用的是第三方库Cheerio的渲染，渲染结果是普通的html结构，
 * 对于snapshot使用render比较合适。
 * 
*/

let props = {
  year: 2021,
  month: 8,
  onChange: jest.fn()
}


let wrapper;

describe('test monthpicker', ()=>{
  beforeEach(()=>{
    wrapper = mount( <MonthPicker {...props}/> )
  })

  it('match snapshot', ()=>{
    expect(wrapper).toMatchSnapshot()
  }) 

  it('render correct year & month, show correct dropdown states', ()=>{
    expect( wrapper.find('.dropdown-toggle').text() ).toEqual(`${ props.year } 年 ${ props.month } 月`)
  }) 
})