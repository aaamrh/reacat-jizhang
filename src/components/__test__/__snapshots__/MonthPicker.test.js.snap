// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test monthpicker match snapshot 1`] = `
initialize {
  "0": Node {
    "attribs": Object {
      "class": "dropdown month-picker-component",
    },
    "children": Array [
      Node {
        "attribs": Object {},
        "children": Array [
          Node {
            "data": "选择月份",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "h4",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Node {
          "attribs": Object {
            "class": "btn btn-secondary dropdown-toggle",
          },
          "children": Array [
            Node {
              "data": "2021 年 8 月",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "text",
            },
          ],
          "name": "button",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {},
        "x-attribsPrefix": Object {},
      },
      Node {
        "attribs": Object {
          "class": "btn btn-secondary dropdown-toggle",
        },
        "children": Array [
          Node {
            "data": "2021 年 8 月",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "button",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Node {
          "attribs": Object {},
          "children": Array [
            Node {
              "data": "选择月份",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "text",
            },
          ],
          "name": "h4",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {},
          "x-attribsPrefix": Object {},
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": Node {
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
    },
    "prev": null,
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;
